packages/valory/skills/staking_abci/behaviours.py

```diff
@@ -0,0 +1,323 @@
+# setting the safe gas to 0 means that all available gas will be used
+# which is what we want in most cases
+# more info here: https://safe-docs.dev.gnosisdev.com/safe/docs/contracts_tx_execution/
+SAFE_GAS = 0
+
+
+class CallCheckpointBehaviour(BaseBehaviour):
+    """Behaviour that calls the checkpoint contract function if the service is staked and if it is necessary."""
+
+    matching_round = CallCheckpointRound
+
+    def __init__(self, **kwargs: Any) -> None:
+        """Initialize the behaviour."""
+        super().__init__(**kwargs)
+        self._is_service_staked: bool = False
+        self._next_checkpoint: int = 0
+        self._checkpoint_data: bytes = b""
+        self._safe_tx_hash: str = ""
+
+    @property
+    def params(self) -> StakingParams:
+        """Return the params."""
+        return cast(StakingParams, self.context.params)
+
+    @property
+    def synchronized_data(self) -> SynchronizedData:
+        """Return the synchronized data."""
+        return SynchronizedData(super().synchronized_data.db)
+
+    @property
+    def synced_timestamp(self) -> int:
+        """Return the synchronized timestamp across the agents."""
+        return int(self.round_sequence.last_round_transition_timestamp.timestamp())
+
+    @property
+    def is_service_staked(self) -> bool:
+        """Whether the service is staked."""
+        return self._is_service_staked
+
+    @is_service_staked.setter
+    def is_service_staked(self, is_service_staked: bool) -> None:
+        """Whether the service is staked."""
+        self._is_service_staked = is_service_staked
+
+    @property
+    def next_checkpoint(self) -> int:
+        """Get the next checkpoint."""
+        return self._next_checkpoint
+
+    @next_checkpoint.setter
+    def next_checkpoint(self, next_checkpoint: int) -> None:
+        """Whether the service is staked."""
+        self._next_checkpoint = next_checkpoint
+
+    @property
+    def is_checkpoint_reached(self) -> bool:
+        """Whether the next checkpoint is reached."""
+        return self.next_checkpoint <= self.synced_timestamp
+
+    @property
+    def checkpoint_data(self) -> bytes:
+        """Get the checkpoint data."""
+        return self._checkpoint_data
+
+    @checkpoint_data.setter
+    def checkpoint_data(self, data: bytes) -> None:
+        """Set the request data."""
+        self._checkpoint_data = data
+
+    @property
+    def safe_tx_hash(self) -> str:
+        """Get the safe_tx_hash."""
+        return self._safe_tx_hash
+
+    @safe_tx_hash.setter
+    def safe_tx_hash(self, safe_hash: str) -> None:
+        """Set the safe_tx_hash."""
+        length = len(safe_hash)
+        if length != TX_HASH_LENGTH:
+            raise ValueError(
+                f"Incorrect length {length} != {TX_HASH_LENGTH} detected "
+                f"when trying to assign a safe transaction hash: {safe_hash}"
+            )
+        self._safe_tx_hash = safe_hash[2:]
+
+    def wait_for_condition_with_sleep(
```

Comment:
Why do we need this?

Reply:
In order to perform the contract calls in a serialized manner, and only continue to the next step if the current call succeeded. Otherwise, the behaviour sleeps and retries the same call.